### PR TITLE
RFC 8037 - Support for Ed25519, Ed448

### PR DESCRIPTION
**Pull request by andrewmackie**
Support for Ed25519 and Ed448, allowing signing and verification.

Partial support for X25519 and X448 (these are now included in JWK but there is no support in JWA wrt shared key exchange).

Some code cherry-picked from [Simo5's rfc8037 branch](https://github.com/simo5/jwcrypto/tree/rfc8037).

No new tests written - given the security implications, I would prefer the tests to be created by an independent and more competent person ...

### CODE REVIEW

**Comment by andrewmackie** [comment]
Sorry, there's a typo in lines 393 and 394 - they're missing a single quote and should read:

            raise InvalidJWKValue('%s is not a supported curve for the ' \
            'OKP key type' % curve)

File: jwcrypto/jwk.py [hunk 1]
```diff
@@ -663,22 +670,37 @@ def _ec_pri(self, k, curve):
...
+        elif k['crv'] == 'X448':
+            return x448.X448PublicKey.from_public_bytes(
+                base64url_decode(k['x']))
+        # No support for other curves
+        raise NotImplementedError
+
+    def _okp_pri(self, k):
+        if k['crv'] == 'Ed25519':
+            return ed25519.Ed25519PrivateKey.from_private_bytes(
+                base64url_decode(k['x']))
```
Comment (simo5) [inline jwcrypto/jwk.py]: these (here and below) should probably be k['d'] ?

File: jwcrypto/jwk.py [hunk 2]
```diff
@@ -688,7 +710,7 @@ def _get_public_key(self, arg=None):
         elif self._params['kty'] == 'EC':
             return self._ec_pub(self._key, arg).public_key(default_backend())
         elif self._params['kty'] == 'OKP':
-            return self._okp_pub(self._key, arg).public_key(default_backend())
+            return self._okp_pri(self._key).public_key()
```
Comment (simo5) [inline jwcrypto/jwk.py]: This will break if you have only a public key in the JWK I think.

File: jwcrypto/jwk.py [hunk 3]
```diff
@@ -332,6 +332,9 @@ def _get_curve_by_name(self, name):
             return ec.SECP384R1()
         elif name == 'P-521':
             return ec.SECP521R1()
+        elif name in ['Ed25519', 'Ed448', 'X25519', 'X448']:
```
Comment (simo5) [inline jwcrypto/jwk.py]: All you need to do here is return the same name you got in input as we need no mapping of names in these cases.

Comment (simo5) [reply jwcrypto/jwk.py]: Another option is to rename this to _get_ec_curve_by_name() and then simply not handle edwards curves in here.
After all this is an internal function and you are not using it in _generate_OKP() anyway.

File: jwcrypto/jwk.py [hunk 4]
```diff
@@ -618,8 +625,8 @@ def get_curve(self, arg):
         :raises InvalidJWKValue: if the curve names is invalid.
         """
         k = self._key
-        if self._params['kty'] != 'EC':
-            raise InvalidJWKType('Not an EC key')
+        if self._params['kty'] not in ['EC', 'OKP']:
+            raise InvalidJWKType('Not an EC or OKP key')
```
Comment (simo5) [inline jwcrypto/jwk.py]: fit the comment above too please.

File: jwcrypto/jwa.py [hunk 5]
```diff
@@ -995,6 +999,28 @@ class _A256Gcm(_AesGcm, JWAAlgorithm):
     algorithm_use = 'enc'
 
 
+class _EdDsa(_RawJWS, JWAAlgorithm):
```
Comment (simo5) [inline jwcrypto/jwa.py]: Let's move this near EcDH

File: jwcrypto/jwa.py [hunk 6]
```diff
@@ -1037,7 +1063,8 @@ class JWA(object):
         'A256CBC-HS512': _A256CbcHs512,
         'A128GCM': _A128Gcm,
         'A192GCM': _A192Gcm,
-        'A256GCM': _A256Gcm
+        'A256GCM': _A256Gcm,
+        'EdDSA': _EdDsa
```
Comment (simo5) [inline jwcrypto/jwa.py]: Same let's move this up near EcDH so asymmetric signature algorithms are close together.

File: jwcrypto/jwk.py (hunk 1 above)
Comment (andrewmackie) [reply jwcrypto/jwk.py]: Yes they should, thank you. I'll fix it.

File: jwcrypto/jwk.py (hunk 2 above)
Comment (andrewmackie) [reply jwcrypto/jwk.py]: Yes. Will fix.

File: jwcrypto/jwk.py (hunk 3 above)
Comment (andrewmackie) [reply jwcrypto/jwk.py]: I'll rename it.

File: jwcrypto/jwk.py (hunk 4 above)
Comment (andrewmackie) [reply jwcrypto/jwk.py]: It now reads:
```
:raises InvalidJWKType: the key is not an EC or OKP key.
```
Is that what you wanted?

File: jwcrypto/jwa.py (hunk 5 above)
Comment (andrewmackie) [reply jwcrypto/jwa.py]: Will do.

File: jwcrypto/jwa.py (hunk 6 above)
Comment (andrewmackie) [reply jwcrypto/jwa.py]: Will do.

File: jwcrypto/jwk.py (hunk 3 above)
Comment (andrewmackie) [reply jwcrypto/jwk.py]: Actually, it's useful for unit testing OKP keys so I'll follow your first suggestion.

**Comment by andrewmackie** [comment]
I've committed the changes and tests requested by Simo5, noting that RFC 8037 only gives an example for Ed25519 (and I don't understand the X25519 and X448 examples there).

I'm happy to add the test data for Ed448, X22519 and X448 if you point me to a source of examples (or other Python libraries) which you trust.

**Comment by simo5** [comment]
@andrewmackie let's dropm X25519 and X448 from your PR, I'll take care of that.
As for tests I am ok with just the Ed25519 check for a first pass, I can add a custom test later if none are defined in the RFC.

File: jwcrypto/jwa.py [hunk 7]
```diff
@@ -693,8 +693,12 @@ def __init__(self):
     def _check_key(self, key):
         if not isinstance(key, JWK):
             raise ValueError('key is not a JWK object')
-        if key.key_type != 'EC':
-            raise InvalidJWEKeyType('EC', key.key_type)
+        if key.key_type not in ['EC', 'OKP']:
+            raise InvalidJWEKeyType('EC or OKP', key.key_type)
+        if key.key_type == 'OKP':
+            if key.key_curve not in ['Ed25519', 'Ed448', 'X25519', 'X448']:
```
Comment (simo5) [inline jwcrypto/jwa.py]: drop X25519 and X448 for now, I will add them later

File: jwcrypto/tests.py [hunk 8]
```diff
@@ -19,14 +19,17 @@
 from jwcrypto.common import base64url_decode, base64url_encode
 from jwcrypto.common import json_decode, json_encode
 
-# RFC 7517 - A.1
+# RFC 7517, RFC 8037 - A.1
```
Comment (simo5) [inline jwcrypto/tests.py]: Please do not mix RFC tests, add a new variable specific for RFC8037 keys unless that RFC reuses keys froma  previous RFC.

**Comment by simo5** [comment]
I think we are close, at this point it would be nice if you could squash everything into a single commit, before applying new changes. Then we'll do a final squash once we are done.

File: jwcrypto/tests.py (hunk 8 above)
Comment (andrewmackie) [reply jwcrypto/tests.py]: Sorry, I'll do that.

File: jwcrypto/jwa.py (hunk 7 above)
Comment (andrewmackie) [reply jwcrypto/jwa.py]: Done

**Comment by andrewmackie** [comment]
Urm ... I didn't close the pull request on purpose. I squashed the commit history down to the last two commits and then force-pushed it:
```
0f4412e (HEAD -> master, github/master) RFC 8037 - Support for Ed25519, Ed448
380601a Changes requested by simo5 + new tests
```
Sorry for being a noob, thanks again for your patience.

**Comment by andrewmackie** [comment]
It seems that the PR [can't be reopened](https://github.com/isaacs/github/issues/361) because my uneducated attempt to squash comments resulted in a 'current head [which] isn't a descendant of the stored head sha'.

So I've added yet another closed PR to your project history, sorry. At this stage I wouldn't blame you if you wanted to take my code and do the PR yourself. 

That said, I can just create a new PR as the repository is ready to go - please let me know.

**Comment by simo5** [comment]
Do not worry, github is a bit asinine with certain rebases, this situation happened to me as well once.

**Comment by simo5** [comment]
Uh, oh seem like you somehow managed to end up squashing the eniter history of the project into a single commit. Once you undo that please either reopen this PR or open a new one.

**Comment by andrewmackie** [comment]
There is no undo for what I did, sorry, so I'll have to start a new PR.
